brainiak/matnorm_models/base.py

```diff
@@ -0,0 +1,229 @@
+            max_abs_current_grad = self.sess.run(max_abs_current_grad_op,
+                                                 feed_dict=feed_dict)
+
+            # check tolerances
+            if abs((current_loss - past_loss) /
+                    max(current_loss, past_loss)) < loss_tol:
+                print('loss tolerance reached on iter %i, %f, stopping' %
+                      (n+1, current_loss))
+                break
+            if max_abs_current_grad < grad_tol:
+                print('gradient tolerance reached on iter %i, %f, stopping' %
+                      (n+1, current_loss))
+                break
+            if (n+1) % step == 0:
+                print('iter %i, %f' % (n+1, current_loss))
+
+    def solve_det_marginal(self, x, sigma, A, Q):
+        """
+        Use matrix inversion lemma for the solve:
+        .. math::
+            (\Sigma + AQA')^{-1} X =\\
+             \Sigma^{-1} - \Sigma^{-1} A (Q^{-1} + A' \Sigma^{-1} A)^{-1} A' \Sigma^{-1}
+
+        Use matrix determinant lemma for determinant:
+        ..math::
+            \log|(\Sigma + AQA')| = \log|Q^{-1} + A' \Sigma^{-1} A| + \log|Q| + \log|\Sigma|
+        """
+
+        # cholesky of (Qinv + A' Sigma^{-1} A)
+        i_qf_cholesky = tf.cholesky(Q.Sigma_inv + tf.matmul(A,
+                                    sigma.Sigma_inv_x(A), transpose_a=True))
+
+        logdet = Q.logdet + sigma.logdet +\
+            2 * tf.reduce_sum(tf.log(tf.matrix_diag_part(i_qf_cholesky)))
+
+        # A' Sigma^{-1}
+        Atrp_Sinv = tf.matmul(A, sigma.Sigma_inv, transpose_a=True)
+        # (Qinv + A' Sigma^{-1} A)^{-1} A' Sigma^{-1}
+        prod_term = tf.cholesky_solve(i_qf_cholesky, Atrp_Sinv)
+        solve = tf.matmul(sigma.Sigma_inv -
+                          sigma.Sigma_inv_x(tf.matmul(A, prod_term)), x)
+        return solve, logdet
+
+    def solve_det_conditional(self, x, sigma, A, Q):
+        """
+        Use matrix inversion lemma for the solve:
+        .. math::
+            (\Sigma - AQA')^{-1} X =\\
+             \Sigma^{-1} + \Sigma^{-1} A (Q^{-1} - A' \Sigma^{-1} A)^{-1} A' \Sigma^{-1}
+
+        Use matrix determinant lemma for determinant:
+        ..math::
+            \log|(\Sigma - AQA')| = \log|Q^{-1} - A' \Sigma^{-1} A| + \log|Q| + \log|\Sigma|
+        """
+
+        # (Qinv - A' Sigma^{-1} A)
+        i_qf_cholesky = tf.cholesky(Q.Sigma - tf.matmul(A,
+                                    sigma.Sigma_inv_x(A), transpose_a=True))
+
+        logdet = -Q.logdet + sigma.logdet +\
+            2 * tf.reduce_sum(tf.log(tf.matrix_diag_part(i_qf_cholesky)))
+
+        # A' Sigma^{-1}
+        Atrp_Sinv = tf.matmul(A, sigma.Sigma_inv, transpose_a=True)
+        # (Q - A' Sigma^{-1} A)^{-1} A' Sigma^{-1}
+        prod_term = tf.cholesky_solve(i_qf_cholesky, Atrp_Sinv)
+
+        solve = tf.matmul(sigma.Sigma_inv +
+                          sigma.Sigma_inv_x(tf.matmul(A, prod_term)), x)
```

Comment:
In general X can be quite large (it is either TRs by voxels or sometimes voxels by TRs). Saving one solve by factoring out Sinv makes sense. The other suggestion (`Prod_x =chol_solve(i_qf_chol, A'*sigma_inv_x(x))`) is dependent on the size of x as you say. By my quick check (below, might be off) if x has many more columns than rows then what I wrote is best, as n approaches m then your suggestion works better. For what we're using this for so far (marginalizing over the TR dimension), voxels >> TRs and therefore the current option should be better, but agreed that conditional structure based on sizes is ideal. Can we register this as an issue for work after the PR? 


```
# X is m*n, Q is k*k, A is m*k, Sigma is m*m

Option 1: 
# A' Sigma^{-1}: km^2 plus the cost of Sigma_inv
Atrp_Sinv = tf.matmul(A, sigma.Sigma_inv, transpose_a=True)
# (Q - A' Sigma^{-1} A)^{-1} A' Sigma^{-1} # k^2 for the solve 
prod_term = tf.cholesky_solve(i_qf_cholesky, Atrp_Sinv)
# m^2k for the inner matmul, m for the addition, m^2 for the solve, m^2n for the outer matmul
solve = tf.matmul(sigma.Sigma_inv_x(scaled_I(1.0, m)+
                          tf.matmul(A, prod_term)), x)

# total: 2km^2 + k^2 + m^2 + nm^2 + m

Option 2:
# m^2 for the solve, kmn for the mult, k^2 for other solve
# (Q - A' Sigma^{-1} A)^{-1} A' Sigma^{-1} X 
prod_term = tf.cholesky_solve(i_qf_cholesky, tf.matmul(A, sigma.Sigma_inv(x), transpose_a = True)
# mkn for inner matmul, m for addition, m^2 for solve
solve = sigma.Sigma_inv_x(scaled_I(1.0, m) + tf.matmul(A, prod_term))

# total: 2kmn + 2m^2 + k^2 + m
```